slashing-protection/src/main/java/tech/pegasys/web3signer/slashingprotection/Pruner.java

```diff
@@ -0,0 +1,127 @@
+    checkArgument(
+        slotsPerEpoch > 0, "slotsPerEpoch must be a positive value, but was %", slotsPerEpoch);
+    final long slotsToKeep = Math.max(epochsToKeep / slotsPerEpoch, 1);
+    pruneBlocks(validatorId, slotsToKeep);
+    pruneAttestations(validatorId, epochsToKeep);
+  }
+
+  private void pruneBlocks(final int validatorId, final long slotsToKeep) {
+    final Optional<UInt64> pruningSlot =
+        jdbi.inTransaction(
+            READ_UNCOMMITTED,
+            h -> {
+              lockForValidator(h, LockType.BLOCK, validatorId);
+              final Optional<UInt64> watermarkSlot =
+                  lowWatermarkDao
+                      .findLowWatermarkForValidator(h, validatorId)
+                      .map(SigningWatermark::getSlot);
+              final Optional<UInt64> slot = signedBlocksDao.findMaxSlot(h, validatorId);
+              final Optional<UInt64> slotToPruneTo =
+                  calculatePruningMark(slotsToKeep, slot, watermarkSlot);
+              slotToPruneTo.ifPresent(
+                  s -> lowWatermarkDao.updateSlotWatermarkFor(h, validatorId, s));
+              return slotToPruneTo;
+            });
+
+    pruningSlot.ifPresent(
+        s ->
+            jdbi.useTransaction(
+                READ_UNCOMMITTED, h -> signedBlocksDao.deleteBlocksBelowSlot(h, validatorId, s)));
+  }
+
+  private void pruneAttestations(final int validatorId, final long epochsToKeep) {
+    final Optional<UInt64> pruningEpoch =
+        jdbi.inTransaction(
+            READ_UNCOMMITTED,
+            h -> {
+              lockForValidator(h, LockType.ATTESTATION, validatorId);
+              final Optional<UInt64> watermarkEpoch =
+                  lowWatermarkDao
+                      .findLowWatermarkForValidator(h, validatorId)
+                      .map(SigningWatermark::getTargetEpoch);
+              final Optional<UInt64> epoch =
+                  signedAttestationsDao.findMaxTargetEpoch(h, validatorId);
+              final Optional<UInt64> epochToPruneTo =
+                  calculatePruningMark(epochsToKeep, epoch, watermarkEpoch);
+              epochToPruneTo.ifPresent(
+                  e -> lowWatermarkDao.updateEpochWatermarksFor(h, validatorId, e, e));
+              return epochToPruneTo;
+            });
+
+    pruningEpoch.ifPresent(
+        e ->
+            jdbi.useTransaction(
+                READ_UNCOMMITTED,
+                h -> signedAttestationsDao.deleteAttestationsBelowEpoch(h, validatorId, e)));
+  }
+
+  private Optional<UInt64> calculatePruningMark(
+      final long amountToPrune,
+      final Optional<UInt64> highpoint,
+      final Optional<UInt64> watermark) {
+    return highpoint.flatMap(
+        h ->
+            watermark.map(
+                w -> {
+                  final UInt64 pruningPoint =
+                      h.compareTo(UInt64.valueOf(amountToPrune)) < 0
+                          ? UInt64.ZERO
+                          : h.subtract(amountToPrune).add(1);
```

Comment:
why add 1?

Reply:
It's due the watermark we are calculating being inclusive of the amount of items to keep.